<div style="
    background: linear-gradient(90deg, #3f007d, #54278f, #6a51a3, #9e9ac8);
    font-family:'JetBrains Mono','Fira Code','Consolas',monospace;
    padding:8px 12px;
    border-radius:10px;
    text-align:center;
    color:#f2f0f7;
    box-shadow:0 4px 14px rgba(0,0,0,0.12);
">
    <h1 style="margin:0; font-size:20px; text-align:center; font-weight:700;">
        📱 Smartphone Price Analysis
    </h1>
        Exploratory Data Analysis
</div>


<div style="
    background:#6a51a3;
    font-family:'JetBrains Mono','Fira Code','Consolas',monospace;
    color:#f2f0f7;
    font-weight:600;
    padding:8px 12px;
    border-left:6px solid #807dba;
    border-radius:6px;
    box-shadow:0 4px 14px rgba(0,0,0,0.12);
    margin-top:4px;
">
    <h1 style="margin:0; font-size:16px; text-align:left; font-weight:700;">
        📚 Library Setup & Initialization
    </h1>
</div>


In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio

import warnings

warnings.filterwarnings('ignore')
# pio.renderers.default = 'iframe'
sns.set(style='darkgrid')
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter('ignore')

<div style="
    background:#6a51a3;
    font-family:'JetBrains Mono','Fira Code','Consolas',monospace;
    color:#f2f0f7;
    font-weight:600;
    padding:8px 12px;
    border-left:6px solid #807dba;
    border-radius:6px;
    box-shadow:0 4px 14px rgba(0,0,0,0.12);
    margin-top:4px;
">
    <h1 style="margin:0; font-size:16px; text-align:left; font-weight:700;">
        🗂️ Load & Initialize Dataset
    </h1>
</div>


In [ ]:
df = pd.read_csv("data/processed/smartphones_cleaned.csv")
pd.set_option("display.max_columns", None)

<div style="
    background:#6a51a3;
    font-family:'JetBrains Mono','Fira Code','Consolas',monospace;
    color:#f2f0f7;
    font-weight:600;
    padding:8px 12px;
    border-left:6px solid #807dba;
    border-radius:6px;
    box-shadow:0 4px 14px rgba(0,0,0,0.12);
    margin-top:4px;
">
    <h1 style="margin:0; font-size:16px; text-align:left; font-weight:700;">
        🧱 Structural & Integrity Checks
    </h1>
</div>


In [3]:
df.head()

smartphone_brand                            model  price_inr  rating_score  \
0        blackzone         BlackZone S25 Ultra Maxx       8999            63   
1           ringme                  Ringme Bold P70       5800            44   
2             lava  Lava Shark 5G (4GB RAM + 128GB)       8998            60   
3             vivo   Vivo Y19s 5G (4GB RAM + 128GB)      11999            69   
4             vivo    Vivo Y19s 5G (4GB RAM + 64GB)      10999            68   

            processor_name processor_brand  core_count  clock_speed_ghz  \
0          unisoc sc9863a1          unisoc           8              1.6   
1           mediatek helio        mediatek           2              1.3   
2              unisoc t765          unisoc           8              2.3   
3  mediatek dimensity 6300        mediatek           8              2.4   
4  mediatek dimensity 6300        mediatek           8              2.4   

   ram_gb  storage_gb  has_5g  has_nfc  has_ir_blaster  display_inches  \
0       8         128   False    False           False            6.60   
1       4         128   False    False           False            6.30   
2       4         128    True    False           False            6.75   
3       4         128    True    False           False            6.74   
4       4          64    True    False           False            6.74   

   res_width_px  res_height_px  refresh_rate_hz  battery_mah  fast_charging  \
0           720           1612               60         5000          False   
1           128            160               60         3260          False   
2           720           1600               90         5000           True   
3           720           1600               90         6000           True   
4           720           1600               90         6000           True   

   charging_watt  rear_camera_count  front_camera_count  rear_camera_main_mp  \
0           10.0                  1                   1                 32.0   
1           10.0                  1                   0                 13.0   
2           10.0                  1                   1                 13.0   
3           15.0                  2                   1                 13.0   
4           15.0                  2                   1                 13.0   

   front_camera_main_mp  os_name memory_card_supported memory_card_type  
0                  16.0  android                   NaN              NaN  
1                   NaN  android                   NaN              NaN  
2                   5.0  android                  True        dedicated  
3                   5.0  android                  True        dedicated  
4                   5.0  android                  True        dedicated

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 758 entries, 0 to 757
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   smartphone_brand       758 non-null    object 
 1   model                  758 non-null    object 
 2   price_inr              758 non-null    int64  
 3   rating_score           758 non-null    int64  
 4   processor_name         758 non-null    object 
 5   processor_brand        758 non-null    object 
 6   core_count             758 non-null    int64  
 7   clock_speed_ghz        758 non-null    float64
 8   ram_gb                 758 non-null    int64  
 9   storage_gb             758 non-null    int64  
 10  has_5g                 758 non-null    bool   
 11  has_nfc                758 non-null    bool   
 12  has_ir_blaster         758 non-null    bool   
 13  display_inches         758 non-null    float64
 14  res_width_px           758 non-null    int64  
 15  res_he

In [5]:
df.describe().round(2)

price_inr  rating_score  core_count  clock_speed_ghz  ram_gb  \
count     758.00        758.00      758.00           758.00  758.00   
mean    29424.09         78.59        7.91             2.60    7.99   
std     32989.69          8.37        0.70             0.60    3.13   
min      4040.00         37.00        1.00             1.00    2.00   
25%     11857.25         74.00        8.00             2.30    6.00   
50%     18999.00         80.00        8.00             2.40    8.00   
75%     29990.00         84.00        8.00             2.80    8.00   
max    229900.00         97.00       10.00             4.47   24.00   

       storage_gb  display_inches  res_width_px  res_height_px  \
count      758.00          758.00        758.00         758.00   
mean       220.50            6.70       1043.92        2234.90   
std        168.37            0.29        260.68         494.16   
min         32.00            3.75        128.00         160.00   
25%        128.00            6.67        720.00        1612.00   
50%        128.00            6.70       1080.00        2400.00   
75%        256.00            6.77       1206.00        2532.00   
max       2048.00            8.03       2400.00        3168.00   

       refresh_rate_hz  battery_mah  charging_watt  rear_camera_count  \
count           758.00       758.00         758.00             758.00   
mean            114.18      5373.57          45.98               2.11   
std              18.06       980.22          28.74               0.71   
min              60.00      2000.00           5.00               1.00   
25%             120.00      5000.00          25.00               2.00   
50%             120.00      5000.00          44.00               2.00   
75%             120.00      6000.00          68.00               3.00   
max             165.00     21200.00         125.00               3.00   

       front_camera_count  rear_camera_main_mp  front_camera_main_mp  
count              758.00               758.00                753.00  
mean                 1.02                52.23                 19.45  
std                  0.17                26.66                 14.48  
min                  0.00                 8.00                  5.00  
25%                  1.00                50.00                  8.00  
50%                  1.00                50.00                 13.00  
75%                  1.00                50.00                 32.00  
max                  2.00               200.00                 50.00

In [6]:
df.isnull().sum()

smartphone_brand           0
model                      0
price_inr                  0
rating_score               0
processor_name             0
processor_brand            0
core_count                 0
clock_speed_ghz            0
ram_gb                     0
storage_gb                 0
has_5g                     0
has_nfc                    0
has_ir_blaster             0
display_inches             0
res_width_px               0
res_height_px              0
refresh_rate_hz            0
battery_mah                0
fast_charging              0
charging_watt              0
rear_camera_count          0
front_camera_count         0
rear_camera_main_mp        0
front_camera_main_mp       5
os_name                    0
memory_card_supported    207
memory_card_type         207
dtype: int64

In [7]:
print("Duplicated Rows:",df.duplicated().sum())
print("-"*30)
print("Number of Rows:",df.shape[0])
print("-"*30)
print("Number of Columns:",df.shape[1])

Duplicated Rows: 0
------------------------------
Number of Rows: 758
------------------------------
Number of Columns: 27


<div style="
    background:#6a51a3;
    font-family:'JetBrains Mono','Fira Code','Consolas',monospace;
    color:#f2f0f7;
    font-weight:600;
    padding:8px 12px;
    border-left:6px solid #807dba;
    border-radius:6px;
    box-shadow:0 4px 14px rgba(0,0,0,0.12);
    margin-top:4px;
">
    <h1 style="margin:0; font-size:16px; text-align:left; font-weight:700;">
        🗂️ Univariate Analysis
    </h1>
</div>


In [8]:
features = {
    "has_5g": "5G",
    "has_nfc": "NFC",
    "has_ir_blaster": "IR Blaster",
    "fast_charging": "Fast Charging",
    "memory_card_supported": "Memory Card",
    "os_name": "Operating System",
}

fig = make_subplots(
    rows=2,
    cols=3,
    specs=[[{"type": "domain"}]*3, [{"type": "domain"}]*3],
    subplot_titles=list(features.values())
)

for idx, (col, title) in enumerate(features.items(), start=1):
    r = 1 if idx <= 3 else 2
    c = idx if idx <= 3 else idx - 3

    fig.add_trace(
        go.Pie(
            labels=df[col].value_counts().index,
            values=df[col].value_counts().values,
            showlegend=False,
            marker=dict(colors=px.colors.sequential.Purp_r),
            hole=0.3,
        ),
        row=r, col=c
    )

fig.update_layout(
    template="plotly_dark",
    title="Feature Availability",
    height=600,
    width=800,
)

# Adjust annotation lift
for ann in fig["layout"]["annotations"]:
    ann["y"] += 0.03

fig.show()

In [9]:
temp_df = df["processor_brand"].value_counts().reset_index()

fig = px.bar(
    temp_df,
    x=temp_df["processor_brand"],
    y=temp_df["count"],
    title="Processor Brand Share",
    color="processor_brand",
    color_discrete_sequence=px.colors.sequential.Purp_r,
    template="plotly_dark",
    height=400,
    width=800,
)

fig.update_layout(showlegend=False)

fig.show()


In [10]:
temp_df=df["smartphone_brand"].value_counts()
brand_count=temp_df[temp_df>=10].reset_index()

fig =  px.bar(
    brand_count,
    x=brand_count["smartphone_brand"],
    y=brand_count["count"],
    title="Smartphone Brand Market Share",
    color="smartphone_brand",
    color_discrete_sequence=px.colors.sequential.Purp_r,
    template="plotly_dark",
    height=600,
    width =800
)
fig.update_layout(showlegend=False)

fig.show()

In [11]:
fig = px.violin(
    df,
    y="clock_speed_ghz",
    points="all",
    title="Clock Speed Distribution",
    color_discrete_sequence=px.colors.sequential.Purp_r,
    template="plotly_dark"
)

fig.update_layout(
    yaxis_title="Clock Speed (GHz)",
    height=800,
    width=800
)

fig.show()

In [12]:
temp_df = df["core_count"].value_counts().reset_index()
temp_df.columns = ["core_count", "count"]

fig = px.pie(
    temp_df,
    names="core_count",
    values="count",
    color_discrete_sequence=px.colors.sequential.Purp_r,
    hole=0.3,
    template="plotly_dark",
    title="Number of Cores Used",
    height=500,
    width=800
)

fig.update_layout(showlegend=False)

fig.show()


In [13]:
fig = px.histogram(
    df,
    x="display_inches",
    color_discrete_sequence=px.colors.sequential.Purp_r,
)

fig.update_layout(
    title="Display Size",
    template="plotly_dark",
    height=600,
    width=800
)

fig.show()


In [14]:
fig = px.histogram(
    df,
    x="battery_mah",
    color_discrete_sequence=px.colors.sequential.Purp_r,
)

fig.update_layout(
    title="Display Size",
    template="plotly_dark",
    height=600,
    width=800
)

fig.show()

In [15]:
plots = {
    "battery_mah": "Battery Capacity Variation Across Smartphones",
    "charging_watt": "Charging Wattage Distribution"
}

colors = px.colors.sequential.Purp_r

fig = make_subplots(
    rows=1, cols=2,
    shared_xaxes=True,
    vertical_spacing=0.05,
    subplot_titles=list(plots.values())
)

for i, (col, title) in enumerate(plots.items(), start=1):
    fig.add_trace(
        go.Violin(
            y=df[col],
            name=title,
            marker=dict(color=colors[i * 2]),
            points="all"
        ),
        row=1, col=i
    )

fig.update_layout(
    template="plotly_dark",
    title="Battery Section Check",
    height=600,
    width=800,
    showlegend=False
)

fig.show()

In [16]:
cams = {
    "front_camera_count": "Front Camera Count Distribution",
    "rear_camera_count": "Rear Camera Count Distribution"
}

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type": "domain"}, {"type": "domain"}]],
    subplot_titles=list(cams.values())
)

for i, (col, title) in enumerate(cams.items(), start=1):
    fig.add_trace(
        go.Pie(
            values=df[col].value_counts().values,
            labels=df[col].value_counts().index,
            hole=0.3,
            showlegend=False,
            marker=dict(colors=px.colors.sequential.Purp_r),
            hovertext=title.replace("Count", "Details"),
            name=title.replace("Count", "Details")
        ),
        row=1, col=i
    )

fig.update_layout(
    template="plotly_dark",
    title="Camera Counts",
    height=600,
    width=800
)

fig.show()

In [17]:
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=["Front Camera Megapixel Distribution", "Rear Camera Megapixel Distribution"]
)

fig.add_trace(go.Histogram(
    y=df["front_camera_main_mp"],
    name="Front Camera",
    orientation="h",
    nbinsy=50,
    marker=dict(color=px.colors.sequential.Purp_r[3]),
),
    row=1, col=1,
)

fig.add_trace(go.Histogram(
    y=df["rear_camera_main_mp"],
    name="Front Camera",
    orientation="h",
    nbinsy=50,
    marker=dict(color=px.colors.sequential.Purp_r[3]),
),
    row=1, col=2,
)
fig.update_layout(
    template="plotly_dark",
    title="Front vs Rear Camera Megapixel Distribution",
    height=600,
    width=800,
)
fig.update_layout(showlegend=False)
fig.show()

In [18]:
plots = {
    "memory_card_supported": "Memory Card Support",
    "memory_card_type": "Supported Card Type"
}

fig = make_subplots(
    rows=1, cols=2,
    specs=[[{"type": "domain"}, {"type": "domain"}]],
    subplot_titles=list(plots.values())
)

for i, (col, title) in enumerate(plots.items(), start=1):
    fig.add_trace(
        go.Pie(
            values=df[col].value_counts().values,
            labels=df[col].value_counts().index,
            hole=0.3,
            showlegend=False,
            marker=dict(colors=px.colors.sequential.Purp_r),
            name=title
        ),
        row=1, col=i
    )

fig.update_layout(
    title="Memory Card Section",
    template="plotly_dark",
    height=600,
    width=800
)

fig.show()

# 🔎 Univariate Analysis – Key Conclusions  
---

- **5G Support**
  - A large portion of the phones already support 5G, showing how quickly the market is shifting toward newer network technologies.
  - This also means 5G may influence price, since it is more common in mid-range and premium devices.

- **NFC Support**
  - NFC is far less common than 5G.
  - It usually appears in higher-priced or more feature-rich phones.
  - Its limited availability makes it a good indicator of whether a device belongs to a mid-range or premium category.

- **IR Blaster**
  - Only a small share of devices offer an IR blaster.
  - This feature is brand-dependent (mainly Xiaomi).
  - It doesn’t strongly relate to pricing across the dataset.

- **Fast Charging**
  - Fast charging is almost universally available, even on budget devices.
  - This shows it has become a standard feature rather than a premium differentiator.

- **Operating System**
  - The dataset is dominated by Android, with very few iOS devices.
  - This means the price distribution is shaped primarily by Android’s mid-range and budget ecosystem.

- **Processor Brand Share**
  - MediaTek is the most widely used processor brand in the dataset, appearing in a large number of devices.
  - Snapdragon holds the second position, commonly seen in mid-range and upper-mid devices.
  - Unisoc and Exynos appear in smaller proportions, mostly in budget and mid-tier phones.
  - Apple and Google Tensor have very few entries, which is expected since iPhones and Pixels make up a small portion of the dataset.
  - Overall, the processor distribution confirms that the dataset is dominated by **Android mid-range hardware**, with only a limited presence of premium flagship chipsets.

- **Smartphones Brand Share**
  - The dataset is dominated by well-known mid-range brands like **Realme, Samsung, Vivo, Oppo, and Xiaomi**, which make up the majority of the entries.
  - Brands such as **Motorola, Lava, Poco, Tecno, iQOO, and Infinix** appear moderately, reflecting their strong presence in the budget and mid-budget segment.
  - Premium or niche brands like **Apple, OnePlus, Nothing, and Google** appear far less frequently.
  - The distribution clearly shows that the dataset is heavily oriented toward **budget and mid-range Android smartphones**, with only a small representation from high-end flagship brands.

- **Clock Speed (GHz)**
  - Most smartphones in the dataset operate within a fairly narrow clock-speed range, typically around **1.8 GHz to 2.4 GHz**.
  - This distribution is expected because modern mid-range processors usually cluster around similar base frequencies.
  - Very high or very low clock speeds are rare, indicating that outliers are minimal for this feature.
  - Overall, clock speed alone does not strongly differentiate devices but works best when combined with other factors like core count and processor brand.

- **Number of Cores Used**
  - The majority of smartphones in the dataset come with **8-core processors**, which aligns with most modern Android devices.
  - The next most common are **6-core processors**, used mainly in **Apple A-series chips** (2 performance + 4 efficiency cores) and a few mid-range Android CPUs.
  - **4-core processors** appear in older or budget smartphones that are still part of the dataset.
  - A very small number of devices use **2-core** or even **1-core** CPUs, indicating extremely low-end or outdated models.
  - **10-core processors** show up only in rare cases (old MediaTek Helio X-series), making them clear outliers.
  - Since 8-core and 6-core devices dominate, core count alone does not fully explain performance — you need to combine it with **processor brand** and **clock speed** for meaningful insights.

- **Display Size (inches)**
  - Most smartphones in the dataset fall in the **6.4 to 6.8 inch** range, which represents today’s standard display size for mid-range devices.
  - A small number of phones have **compact screens around 5.4–5.8 inches**, indicating older or rare small-form-factor models.
  - A few devices extend beyond **7 inches**, entering the phablet/tablet-like category (e.g., foldables or extra-large displays).
  - The overall distribution is heavily concentrated toward larger screens, showing that modern smartphones increasingly favor big displays.
  - Display size on its own doesn’t strongly separate price categories, but when combined with **resolution** and **refresh rate**, it becomes more meaningful for differentiating budget vs premium models.

- **Battery Capacity (mAh)**
  - Most smartphones in the dataset have a battery capacity between **4500 mAh and 6000 mAh**, with **5000 mAh** being the most common.
  - A few devices include larger batteries in the **6500–7000 mAh** range, usually meant for heavy-use or rugged models.
  - There is one **extreme outlier at 21,200 mAh**, which is far beyond normal smartphone capacity.  
    This value will distort analysis and should be removed before model training.
  - A small number of older or budget phones fall below **3000 mAh**, forming a low-end outlier group.
  - Overall, the battery distribution is **right-skewed**, with most phones clustering around typical modern capacities.

- **Charging Wattage**
  - Most phones offer **18W to 33W** fast charging, which has become the industry standard.
  - Higher wattage options like **44W, 50W, 67W, and 120W** appear in mid-range and premium phones.
  - A few models reach extremely high charging speeds, but these are rare compared to the rest of the dataset.
  - Charging wattage shows a wider spread than battery size, but still follows a clear pattern where higher wattage is found in higher-priced devices.

- **Front Camera Count**
  - Most smartphones in the dataset come with a **single front camera**, which is the standard setup across the industry.
  - A smaller number of devices use **dual front cameras**, usually found in selfie-focused or premium models.
  - Front camera counts are generally simple and do not vary much, so they contribute only limited differentiation among devices.

- **Rear Camera Count**
  - The majority of smartphones feature **triple-camera or quad-camera** setups on the rear.
  - Dual-camera phones appear in moderate numbers, typically in older or budget models.
  - Single-camera devices are very rare and mostly belong to older generations.
  - Overall, rear camera count shows more variety than the front camera count and can help separate budget phones (2 cameras) from mid-range and premium phones (3–4 cameras).

- **Front Camera Megapixel Distribution**
  - Most smartphones have front cameras in the **8 MP to 16 MP** range, which is the standard for mid-range devices.
  - A few models offer **20 MP or higher** front cameras, usually targeting users who prioritize selfies.
  - Extremely low front MP values appear in a small number of older or entry-level phones.
  - Overall, the distribution is fairly concentrated, meaning front MP does not vary wildly across phones.

- **Rear Camera Megapixel Distribution**
  - Rear camera megapixels show a **much wider spread** compared to front cameras.
  - Common values include **12 MP, 48 MP, 50 MP, and 64 MP**, which are popular across many brands.
  - Some models reach very high MP counts like **108 MP** or **200 MP**, which are typical of premium or camera-centric phones.
  - The distribution has clear peaks at marketing-driven megapixel values (e.g., “48MP”, “64MP”, “108MP”).

- **Overall Comparison(Front Camera Megapixel Distribution and Rear Camera Megapixel Distribution)**
  - Rear cameras show **far more variation** than front cameras, both in range and in peak values.
  - Higher MP counts on the rear side often indicate **better camera hardware** and may correlate with higher prices.
  - Front cameras are more standardized, and megapixel differences are less dramatic, making them a weaker indicator of pricing.

- **Memory Card Support**
  - A noticeable portion of smartphones **do not support memory cards**, especially newer mid-range and premium models.
  - Phones that do support memory expansion are mostly from budget and mid-budget segments.
  - The absence of a memory card slot often indicates a more modern or premium device design.
  - This feature may help distinguish lower-priced devices from higher-priced ones.

- **Memory Card Type**
  - Among the phones that support external storage, **microSD** is the dominant and nearly universal card type.
  - Very few models, if any, use alternative card formats.
  - Since type only matters when support is available, this feature adds limited additional value compared to the support flag itself.

- **Overall Observation(Memory Card Support and Memory Card Type)**
  - Memory card support is becoming less common in newer smartphones.
  - Devices without memory card support tend to fall into the **mid-range and premium categories**, which may impact pricing.
  - Memory card type has little variation and should be used only alongside the support indicator.


# 🔄 Bivariate Analysis Coming Soon

Univariate analysis is done for now — I’ll add the bivariate part soon with price comparisons and more visuals.

If this notebook or dataset helped you, a quick **upvote** would really motivate me to keep improving it. Thanks for the support! ❤️🚀
